File: app/backend/src/main/java/io/tubs/app/service/CompanyService.java
package io.tubs.app.service;

import io.tubs.app.dto.CompanyDto;
import io.tubs.app.dto.RegistrationRequest;
import io.tubs.app.exception.NotFoundException;
import io.tubs.app.mapper.CompanyMapper;
import io.tubs.app.model.Company;
import io.tubs.app.repository.CompanyRepository;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.stereotype.Service;

@Service
@Slf4j
@RequiredArgsConstructor
public class CompanyService {

    private final CompanyRepository companyRepository;

    public void register(RegistrationRequest request) {
        Company account = new Company(
                request.companyNumber(),
                request.companyName(),
                request.directorName(),
                request.directorEmail(),
                request.city(),
                request.postalCode(),
                request.street(),
                request.houseNumber()
        );
        companyRepository.save(account);
    }

    public CompanyDto get(String companyNumber) {
        Company company = companyRepository.findByCompanyNumber(companyNumber).orElseThrow(() -> new NotFoundException("Company does not exist"));
        return CompanyMapper.toDto(company);
    }

    public CompanyDto update(CompanyDto companyDto) {
        Company company = companyRepository.findByCompanyNumber(companyDto.companyNumber()).orElseThrow(() -> new NotFoundException("Company does not exist"));
        company.setPaymentAccountName(companyDto.paymentAccountName());
        company.setPaymentTerms(companyDto.paymentTerms());
        company.setIban(companyDto.iban());
        company.getRegisteredOffice().setCity(companyDto.registeredOffice().city());
        company.getRegisteredOffice().setPostalCode(companyDto.registeredOffice().postalCode());
        company.getRegisteredOffice().setStreet(companyDto.registeredOffice().street());
        company.getRegisteredOffice().setHouseNumber(companyDto.registeredOffice().houseNumber());
        companyRepository.save(company);
        return CompanyMapper.toDto(company);
    }

    public void unregister(String companyNumber) {
        Company company = companyRepository.findByCompanyNumber(companyNumber).orElseThrow(() -> new NotFoundException("Company does not exist"));
        company.setRegisteredOnPeppol(false);
        companyRepository.save(company);
    }
}

File: app/ui/src/services/app/company-service.ts
import {resolve} from '@aurelia/kernel';
import {singleton} from "aurelia";
import {AppApi} from "./app-api";

export interface CompanyDto {
    name: string,
    companyNumber: string,
    subscriber: string,
    subscriberEmail: string,
    paymentTerms: string,
    iban: string,
    paymentAccountName: string,
    registeredOnPeppol: boolean,
    registeredOffice: Address
}

export interface Address {
    city?: string,
    postalCode?: string,
    street?: string,
    houseNumber?: string
}

@singleton()
export class CompanyService {
    private appApi = resolve(AppApi);
    public myCompany: CompanyDto;

    async getAndSetMyCompanyForToken() : Promise<CompanyDto> {
        this.myCompany = await this.appApi.httpClient.get(`/api/company`).then(response => response.json());
        return Promise.resolve(this.myCompany);
    }

    async updateCompany(company: CompanyDto) {
        const response = await this.appApi.httpClient.put(`/api/company`, JSON.stringify(company) );
        this.myCompany = await response.json();
        return Promise.resolve(this.myCompany);
    }

}
